### PR TITLE
New resource to create an environment in octopus

### PR DESCRIPTION
**Pull request by matt-richardson**

### CODE REVIEW

**Comment by matt-richardson** [comment]
Ready for proper review now, @stopthatastronaut.

File: OctopusDSC/DSCResources/cOctopusEnvironment/cOctopusEnvironment.psm1 [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  $tempFolder = [System.IO.Path]::GetTempPath()
+  $shadowCopyFolder = Join-Path $tempFolder ([Guid]::NewGuid())
+  New-Item -type Directory $shadowCopyFolder | Out-Null
+
+  $filename = "${env:ProgramFiles}\Octopus Deploy\Octopus\Newtonsoft.Json.dll"
+  $version = [System.Diagnostics.FileVersionInfo]::GetVersionInfo($filename)
+  Write-Verbose "Shadow copying '$filename' (version $($version.FileVersion)) to $shadowCopyFolder"
+  Copy-Item $filename $shadowCopyFolder
+
+  $filename = "${env:ProgramFiles}\Octopus Deploy\Octopus\Octopus.Client.dll"
```
Comment (stopthatastronaut) [inline OctopusDSC/DSCResources/cOctopusEnvironment/cOctopusEnvironment.psm1]: Pretty sure we could do this without Octopus.Client (doesn't affect whether this PR gets merged, obvs, but I'd lean towards fewest dependencies wherver we can)

File: OctopusDSC/DSCResources/cOctopusEnvironment/cOctopusEnvironment.psm1 [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    $credentials = New-Object Octopus.Client.Model.LoginCommand
+    $credentials.Username = $OctopusCredentials.GetNetworkCredential().Username
+    $credentials.Password = $OctopusCredentials.GetNetworkCredential().Password
+    $repository.Users.SignIn($credentials)
+  }
+
+  return $repository
+}
+
+function Get-OctopusDSCParameter($parameters) {
```
Comment (stopthatastronaut) [inline OctopusDSC/DSCResources/cOctopusEnvironment/cOctopusEnvironment.psm1]: As for previous reviews, we really need to have a shared module for this stuff - which is duplicated across multiple resources

File: OctopusDSC/Examples/cOctopusEnvironment.ps1 [hunk 3]
```diff
@@ -0,0 +1,18 @@
+Configuration SampleConfig
+{
+    Import-DscResource -Module OctopusDSC
+
+    Node "localhost"
+    {
+        $pass = ConvertTo-SecureString "Password" -AsPlainText -Force
```
Comment (stopthatastronaut) [inline OctopusDSC/Examples/cOctopusEnvironment.ps1]: I suspect this'll fail the Gallery's script analyzer again. They do scan examples. The cOctopusServer example has a different method in it now, using a pre-encrypted password file.

File: OctopusDSC/Tests/cOctopusEnvironment.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                    $result.Ensure | should be 'Absent'
+                }
+            }
+
+            Context 'Test-TargetResource' {
+                $response = @{  }
+                Mock Get-TargetResource { return $response } -Scope It
+
+                It 'Returns false if environment does not exist' {
+                    $desiredPassword = ConvertTo-SecureString "1a2b3c4d5e6f" -AsPlainText -Force
```
Comment (stopthatastronaut) [inline OctopusDSC/Tests/cOctopusEnvironment.Tests.ps1]: Will this fail the Gallery tests again? Not sure if Tests get excluded

File: OctopusDSC/DSCResources/cOctopusEnvironment/cOctopusEnvironment.psm1 (hunk 1 above)
Comment (matt-richardson) [reply OctopusDSC/DSCResources/cOctopusEnvironment/cOctopusEnvironment.psm1]: The only problem with that is we'd have to implement the OctopusClients auth model to support username/password, which is non-trivial.

**Pull request merged**